utilities/qneg.F90

```diff
@@ -0,0 +1,381 @@
+  public :: qneg_timestep_final
+  public :: qneg_final
+
+  ! Private module variables
+  logical          :: collect_stats = .false.
+  logical          :: timestep_reset = .false.
+  logical          :: qneg_initialized = .false.
+  integer          :: num_constituents = 0
+
+  real(kind_phys), parameter :: tol = 1.e-12_kind_phys
+  real(kind_phys), parameter :: worst_reset = 1.e35_kind_phys
+
+  ! Diagnostic field names
+!  integer              :: num_diag_fields = 0
+!  character(len=max_fieldname_len) :: diag_names(num_diag_fields)
+!  logical             :: cnst_out_calc = .false.
+!  logical             :: cnst_outfld(num_diag_fields) = .false.
+
+  ! Summary buffers
+  integer, parameter    :: num_bins = 24
+  character(len=CS)     :: qneg_warn_labels(num_bins) = ''
+  integer, allocatable  :: qneg_warn_num(:,:)
+  real(kind_phys), allocatable :: qneg_warn_worst(:,:)
+
+  private             :: qneg_print_summary
+!  private             :: calc_cnst_out
+  private             :: find_index
+  interface reset_stats
+     module procedure reset_stats_scalar
+     module procedure reset_stats_array
+  end interface reset_stats
+
+contains
+
+!> \section arg_table_qneg_init Argument Table
+!! \htmlinclude qneg_init.html
+  subroutine qneg_init(print_qneg_warn, number_of_constituents, qmin, errcode, errmsg)
+    !use cam_history,    only: addfld, horiz_only
+    !use constituents,   only: cnst_longname
+
+    character(len=*),  intent(in)  :: print_qneg_warn
+    integer,           intent(in)  :: number_of_constituents
+    real(kind_phys),   intent(out) :: qmin(:)
+    integer,           intent(out) :: errcode
+    character(len=512),intent(out) :: errmsg
+
+    character(len=*), parameter  :: subname = 'qneg_init'
+    integer                      :: ierr
+
+    errcode = 0
+    errmsg = ''
+
+    !Check if already initialized:
+    if (qneg_initialized) then
+       return
+    end if
+
+!    integer :: index
+
+!    do index = 1, num_constituents
+!       diag_names(index) = trim(cnst_name(index))//'_qneg'
+!       call addfld(diag_names(index), (/ 'lev' /), 'I', 'kg/kg',              &
+!            trim(cnst_longname(index))//' QNEG error (cell)')
+!       diag_names(num_constituents+index) = trim(cnst_name(index))//'_qneg_col'
+!       call addfld(diag_names(num_constituents+index), horiz_only, 'I', 'kg/kg',         &
+!            trim(cnst_longname(index))//' QNEG error (column)')
+!    end do
+!    diag_names((2*num_constituents) + 1) = 'qflux_exceeded'
+!    call addfld(diag_names((2*num_constituents) + 1), horiz_only, 'I', 'kg/m^2/s',     &
+!         'qflux excess (QNEG4)')
+
+    !Create fake qmin -> will be removed when qmin exists
+    qmin = 0._kind_phys
+
+    !Allocate and initialize arrays whose dimensions depend on num_constituents:
+    num_constituents = number_of_constituents
+    allocate(qneg_warn_num(num_constituents, num_bins), stat=ierr)
+    call check_allocate(ierr, subname, 'qneg_warn_num')
```

Comment:
It might be worth discussing at some point how CCPP schemes can raise an error associated with allocation failures without having to bring-in the `check_allocate` function via a CAMDEN-specific `use` statement.  Maybe add this to the issue I mentioned above as a discussion topic?

Reply:
Oh, good point. I am so used to seeing the `check_allocate` calls that I forgot we were in a CCPP scheme. These should be using `errcode` and `errmsg` followed by `return`.